scala/private/rules/scala_doc.bzl

```diff
@@ -82,18 +82,27 @@ def _scala_doc_impl(ctx):
 
     # Construct scaladoc args, which also include scalac args.
     # See `scaladoc -help` for more information.
+    argfile = ctx.actions.declare_file("scaladoc_args.txt")
+
+    # turn off warnings for now since they can obscure actual errors for large scala_doc targets
+    argfile_contents = ["-usejavacp", "-nowarn"]
+    argfile_contents.extend(ctx.attr.scalacopts)
+    argfile_contents.append("-d {}".format(output_path.path))
+    argfile_contents.extend(["-Xplugin:{}".format(plugin) for plugin in plugins.to_list()])
+    argfile_contents.append("-classpath {}".format(
+        ctx.configuration.host_path_separator.join([c.path for c in classpath.to_list()]),
+    ))
+    argfile_contents.extend([src_file.path for src_file in src_files.to_list()])
+    ctx.actions.write(
+        output = argfile,
+        content = "\n".join(argfile_contents),
+    )
     args = ctx.actions.args()
-    args.add("-usejavacp")
-    args.add("-nowarn")  # turn off warnings for now since they can obscure actual errors for large scala_doc targets
-    args.add_all(ctx.attr.scalacopts)
-    args.add("-d", output_path.path)
-    args.add_all(plugins, format_each = "-Xplugin:%s")
-    args.add_joined("-classpath", classpath, join_with = ctx.configuration.host_path_separator)
-    args.add_all(src_files)
+    args.add("@{}".format(argfile.path))
```

Comment:
It think there should be enough just to add the following to the args
```
args.set_param_file_format("multiline")
args.use_param_file(param_file_arg = "@%s", use_always = True)
```
and there is no need to declare a file and write content to it. Bazel should write args file automatically when it's passed as an argument to the action.

Example is here https://github.com/bazelbuild/rules_scala/blob/master/scala/private/rule_impls.bzl#L74-L75

Reply:
Thanks for the help, makes life much easier!